### PR TITLE
doc(wyh):MAAC doc

### PR DESCRIPTION
**Pull request by Weiyuhong-1998**
MAAC framework doc

### CODE REVIEW

File: source/best_practice/maac.rst [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        x = self.critic_head(x)
+        return {'value': x['pred']}
+
+Policy
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When modifying the single-agent algorithm into a multi-agent algorithm, the policy part basically remains the same, the only thing to note is to add the multi_agent key and to call the multi-agent model when the multi_agent key is True
+
+.. code:: 
+
+    def default_model(self) -> Tuple[str, List[str]]:
```
Comment (puyuan1996) [inline source/best_practice/maac.rst:47]: 是否可以再加上masac的简单介绍？还是这里只需要讲mappo
 def default_model(self) -> Tuple[str, List[str]]:
        return 'maqac', ['ding.model.template.maqac']

File: source/best_practice/maac.rst [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        logit[action_mask == 0.0] = -99999999
+        return {'logit': logit}
+    def compute_critic(self, x: Dict) -> Dict:
+        x = self.critic_encoder(x['global_state'])
+        x = self.critic_head(x)
+        return {'value': x['pred']}
+
+Policy
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When modifying the single-agent algorithm into a multi-agent algorithm, the policy part basically remains the same, the only thing to note is to add the multi_agent key and to call the multi-agent model when the multi_agent key is True
```
Comment (puyuan1996) [inline source/best_practice/maac.rst:43]: 句尾句号

File: source/best_practice/maac.rst [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+         model=dict(
+               # (int) agent_num: The number of the agent.
+               # For SMAC 3s5z, agent_num=8; for 2c_vs_64zg, agent_num=2.
+               agent_num=agent_num,
+               # (int) obs_shape: The shapeension of observation of each agent.
+               # For 3s5z, obs_shape=150; for 2c_vs_64zg, agent_num=404.
+               # (int) global_obs_shape: The shapeension of global observation.
+               # For 3s5z, obs_shape=216; for 2c_vs_64zg, agent_num=342.
+               agent_obs_shape=72,
+               #global_obs_shape=216,
```
Comment (puyuan1996) [inline source/best_practice/maac.rst:96]: 可以给不同环境都加上相应的shape参数，方便读者使用

File: source/best_practice/maac.rst [hunk 4]
```diff
@@ -0,0 +1,289 @@
+Multi-Agent Actor-Critic RL
+============================================
+MARL algorithms can be divided into two broad categories: centralized learning and decentralized learning. Recent work has developed two lines of research to bridge the gap betewwn these two frameworks: centralized training and decentralized execution(CTDE) and value decomposition(VD).
```
Comment (Will-Nie) [inline source/best_practice/maac.rst:3]: 'between' instead of 'betewwn'

File: source/best_practice/maac.rst [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+VD such as Qmix typically represents the joint Q-function as a function of agents’ local Q-functions, which has been considered as the gold standard for many MARL benchmarks.
+CTDE methods such as MADDPG, MAPPO and COMA improve upon decentralized RL by adopting an actor-critic structure and learning a centralized critic. 
+In DI-engine, we introduced the multi-agent actor-critic framework to quickly convert a single-agent algorithm into a multi-agent algorithm.
```
Comment (Will-Nie) [inline source/best_practice/maac.rst:6]: `introduce` instead of `introduced`

File: source/best_practice/maac.rst [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+
+For User
```
Comment (Will-Nie) [inline source/best_practice/maac.rst:9]: For `Users` instead of `User`

File: source/best_practice/maac.rst [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+.. code::python 
+
+   return {
+         'agent_state': torch.randn(B, agent_num, agent_obs_shape),
+         'global_state': torch.randn(B, agent_num, global_obs_shape),
+         'action_mask': torch.randint(0, 2, size=(B, agent_num, action_shape))
+   }
+
+- agent state: Agent state is ecah agent's local observation.
```
Comment (Will-Nie) [inline source/best_practice/maac.rst:24]: An agent state comprises of each agent's local observation.

File: source/best_practice/maac.rst [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+- global state: Global state contains all global information that can't be seen by each agents.
```
Comment (Will-Nie) [inline source/best_practice/maac.rst:25]: A global state contains......cannot be seen by each agents .

File: source/best_practice/maac.rst [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+   - 3s5z + MAPPO/IPPO
+   .. image:: images/3s5z_mappo.png
+     :align: center
+
+For Developer
+--------------------------
+
+Model
+^^^^^^^^^^^^^^^^^^
+We need to change the signal agent to the multi agent model. In signal agent model, it only has a obs_shape key. In multi agent model, we need to divide the obs_shape key to agent_obs_shape and global_obs_shape, and in this way, we can train critic model by global obs and train actor model by agent obs.
```
Comment (puyuan1996) [inline source/best_practice/maac.rst:196]: signal -> single

File: source/best_practice/maac.rst [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+        adv = torch.zeros_like(value, device=value.device)
+        gae_item = torch.zeros_like(value[0])
+
+        for t in reversed(range(reward.shape[0])):
+            gae_item = delta[t] + factor * gae_item * (1 - done[t])
+            adv[t] += gae_item
+        return adv
+
+When we change your codes, we need to test our codes by the following way.
+You can just input (B, N) data to test signal agent rl utils codes and input (B, A, N) data to test multi agent rl utils codes.
```
Comment (puyuan1996) [inline source/best_practice/maac.rst:251]: single

File: source/best_practice/maac.rst [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+        Overview:
+            Implementation of Generalized Advantage Estimator
+        """
+        value, next_value, reward, done = data
+        if done is None:
+            done = torch.zeros_like(reward, device=reward.device)
+        # In Multi-agent RL, the value's dimension is (B, A, N), but the reward's dimension is (B, N)
+        if len(value.shape) == len(reward.shape) + 1:
+            reward = reward.unsqueeze(-1)
+            done = done.unsqueeze(-1)
```
Comment (puyuan1996) [inline source/best_practice/maac.rst:239]: the reward's dimension is (B, N)?

File: source/best_practice/maac.rst [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+   action_shape = 14
+   return {
+         'agent_state': torch.randn(agent_num, agent_obs_shape),
+         'global_state': torch.randn(agent_num, global_obs_shape),
+         'action_mask': torch.randint(0, 2, size=(agent_num, action_shape))
+   }
+
+- agent state: Agent state is ecah agent's local observation.
+- global state: Global state contains all global information that can't be seen by each agents.
+- action Mask: In multi-agent games, it is often the case that some actions cannot be executed due to game constraints. For example, in SMAC, an agent may have skills that cannot be performed frequently. So, when computing the logits for the softmax action probability, we mask out the unavailable actions in both the forward and backward pass so that the probabilities for unavailable actions are always zero. We find that this substantially accelerates training. The data type is \ ``int``\.
```
Comment (puyuan1996) [inline source/best_practice/maac.rst:30]: should add the relative reference like mappo paper for these tricks.

File: source/best_practice/maac.rst [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+- independent obs: The global observation is as same as agent observation, we use it in independent PPO, independent SAC alg. If you want to use it, you have to set ``independent_obs`` to ``True`` in env config.
+
+Model
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+- Centralized training and decentralized executed: Unlike single-agent environments that feed the same observation information to actor and critic networks, in multi-agent environments, we feed agent_state and action_mask information to the actor network to get each actions' logits and mask the invalid/inaccessible actions. At the same time, we feed global_state information to the critic network to gei the global critic value.
+- Action mask: We need to mask the invalid/inaccessible actions when we train or collect data. So we use ``logit[action_mask == 0.0] = -99999999`` to make the inaccessible actions' probability to a very low value. So we can't choose this action when we collect data or train the model. If you don't want to use it, just delect ``logit[action_mask == 0.0] = -99999999``.
```
Comment (Will-Nie) [inline source/best_practice/maac.rst:43]: `delete` instead of `delect`
